### PR TITLE
Abstracted Join to Custom Component

### LINKED ISSUE #1467: Abstract to Custom Component
image

- To better align with the rest of the file structure, put the join contents into a separate component which is then called

### LINKED ISSUE #1468: Use Async/Await
image

- Code readability takes a hit when its async, need to make it sync since it has no performance benefit but makes it easier to understand

### LINKED ISSUE #1469: User Async/Await
image

 - Code readability takes a hit when its async, need to make it sync since it has no performance benefit but makes it easier to understand

### PR DESCRIPTION
**Pull request by Fuyuki325**
idk what the restrictions parameter is but the page works

### CODE REVIEW

File: src/app/user/join/[teamID]/page.js [hunk 1]
```diff
@@ -1,83 +1,20 @@
...
-                {member.email}
-              </span>
-            </p>
-          ))}
-          <Button color="green" size="xl" text="Join" onClick={handleJoin} />
-        </div>
-      )}
-    </div>
+    <ProtectedPage
+      title={`User | Join | ${capitalizeFirstLetter(params.teamID)}`}
```
Comment (shahdivyank) [inline src/app/user/join/[teamID]/page.js]: only need User | Join for the title

File: src/app/user/join/[teamID]/page.js [hunk 2]
```diff
@@ -1,83 +1,20 @@
...
+      restrictions={{
+        participants: [-1, 0, 1],
```
Comment (shahdivyank) [inline src/app/user/join/[teamID]/page.js]: rejected participants cannot join teams, only accepted

File: src/components/user/join/Join.jsx [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+      sessionUpdate({
+        team: teamID,
+      });
+      router.push("/user");
+    });
+  };
+
+  useEffect(() => {
+    if (teamID) {
+      api({
```
Comment (shahdivyank) [inline src/components/user/join/Join.jsx]: use await/async

File: src/components/user/join/Join.jsx [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+
+const Join = ({ params }) => {
+  const [team, setTeam] = useState(null);
+  const router = useRouter();
+  const { update: sessionUpdate } = useSession();
+
+  const { teamID } = params;
+
+  const handleJoin = () => {
+    api({
```
Comment (shahdivyank) [inline src/components/user/join/Join.jsx]: use await/async

File: src/components/user/join/Join.jsx [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+      }
+      toaster("Successfully joined team!", "success");
+      sessionUpdate({
+        team: teamID,
+      });
+      router.push("/user");
+    });
+  };
+
+  useEffect(() => {
```
Comment (shahdivyank) [inline src/components/user/join/Join.jsx]: use effect can be converted into a server component, make the DB call within the component and pass it into a client component for any interaction based things

File: src/components/user/join/Join.jsx [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+import { useSession } from "next-auth/react";
+import Fault from "@/utils/error";
+import { api } from "@/utils/api";
+
+const Join = ({ params }) => {
+  const [team, setTeam] = useState(null);
+  const router = useRouter();
+  const { update: sessionUpdate } = useSession();
+
+  const { teamID } = params;
```
Comment (shahdivyank) [inline src/components/user/join/Join.jsx]: call it id

**Comment by Fuyuki325** [comment]
i tested with a new accepted account and added a new member. i tested with a non-accepted account and was denied access as expected. seems to work unless there is some edge case i cannot come up with

File: src/actions/join.js [hunk 7]
```diff
@@ -0,0 +1,14 @@
+// import Fault from "@/utils/error";
```
Comment (shahdivyank) [inline src/actions/join.js]: remove comments

File: src/actions/join.js [hunk 8]
```diff
@@ -0,0 +1,14 @@
...
+import { db } from "@/utils/firebase";
+import { doc, getDoc } from "firebase/firestore";
+import Fault from "@/utils/error";
+
+export async function fetchTeamData(id) {
```
Comment (shahdivyank) [inline src/actions/join.js]: call it just fetchTeam

File: src/actions/join.js [hunk 9]
```diff
@@ -0,0 +1,14 @@
...
+  const docRef = doc(db, "teams", id);
+  const docSnap = await getDoc(docRef);
```
Comment (shahdivyank) [inline src/actions/join.js]: combine this line and the line above

File: src/components/user/join/Join.jsx [hunk 10]
```diff
@@ -0,0 +1,11 @@
+import { fetchTeamData } from "@/actions/join";
+import JoinClient from "@/components/user/join/JoinClient";
+
+const Join = async ({ params }) => {
+  const { teamID: id } = params;
+  const team = await fetchTeamData(id);
+
+  return <div>{team ? <JoinClient team={team} id={id} /> : ""}</div>;
```
Comment (shahdivyank) [inline src/components/user/join/Join.jsx]: should just be team && <Join

File: src/components/user/join/JoinClient.jsx [hunk 11]
```diff
@@ -0,0 +1,54 @@
...
+const JoinClient = ({ team, id }) => {
+  const router = useRouter();
+  const { update: sessionUpdate } = useSession();
+
+  const handleJoin = async () => {
+    await api({
+      method: "PUT",
+      url: "/api/members",
+      body: { team: id },
+    }).then((response) => {
```
Comment (shahdivyank) [inline src/components/user/join/JoinClient.jsx]: .then and .catch get removed when u do await

**Comment by shahdivyank** [comment]
Attach the other 2 issues as well

File: src/components/user/join/JoinClient.jsx [hunk 12]
```diff
@@ -0,0 +1,53 @@
...
+    }
+    toaster("Successfully joined team!", "success");
+    sessionUpdate({
+      team: id,
+    });
+    router.push("/user");
+  };
+
+  return (
+    <div>
```
Comment (shahdivyank) [inline src/components/user/join/JoinClient.jsx]: react fragment

**Review by shahdivyank — APPROVED** [review]

**Pull request merged**